cylc/flow/cfgspec/suite.py

```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
+                                      |
+                                      |    +-------------------------------+
+                                      +----> Return without changes        |
+                                           +-------------------------------+
+
+    Args (cfg):
+        config object to be upgraded
+
+    Returns (cfg):
+        upgraded config object
+    """
+    # If platform and old settings are set fail
+    # and remote should be added to this forbidden list
+    forbidden_with_platform = [
+        'host', 'batch system', 'batch submit command template'
+    ]
+
+    for task_name, task_spec in cfg['runtime'].items():
+        # if task_name == 'delta':
+        #     breakpoint(header=f"task_name = {task_name}")
+
+        if (
+            'platform' in task_spec and 'job' in task_spec or
+            'platform' in task_spec and 'remote' in task_spec
+        ):
+            if (
+                'platform' in task_spec and
+                any(forbidden_with_platform == task_spec['job']) or
+                any(forbidden_with_platform == task_spec['remote'])
+            ):
+                # Fail Loudly and Horribly
+                LOG.error(
+                    f"A mixture of Cylc 7 (host) and Cylc 8 (platform logic)"
+                    f" should not be used. Task {task_name} set platform "
+                    f"and item in {forbidden_with_platform}"
+                )
+        elif 'platform' in task_spec:
+            # Return config unchanged
+            continue
+
+        else:
+            # Add empty dicts if appropriate sections not present.
+            if 'job' in task_spec:
+                task_spec_job = task_spec['job']
+            else:
+                task_spec_job = {}
+            if 'remote' in task_spec:
+                task_spec_remote = task_spec['remote']
+            else:
+                task_spec_remote = {}
+
+            # Deal with case where host is a function and we cannot auto
+            # upgrade at the time of loading the config.
+            if (
+                'host' in task_spec_remote and
+                REC_COMMAND.match(task_spec['remote']['host'])
+            ):
+                LOG.info(
+                    f"Unable to upgrade task '{task_name}' to platform at "
+                    f"validation because the host setting contains a "
+                    f"function. Cylc will attempt to upgrade this task on"
+                    f" job submission."
+                )
+                continue
+
+            # Attempt to use the reverse lookup
+            try:
+                platform = reverse_lookup(
+                    glbl_cfg(cached=False).get(['job platforms']),
+                    task_spec_job,
+                    task_spec_remote
+                )
+            except PlatformLookupError as exc:
+                LOG.error(f'Unable to determine platform for {task_name}')
+                LOG.debug(f'Exception was {exc}')
```

Comment:
If we can't determine a platform we can't run the suite so I guess just let this error float to the top or re-raise it as a `ConfigError` with more helpful text.

Reply:
Do you mean use Python exceptions rather than the Cylc Logging?

Reply:
Yes, if we don't fail loudly here it will become a runtime error later on.

Reply:
Soundeth Good. Will do.